runtime/lua/vim/keymap.lua

```diff
@@ -0,0 +1,136 @@
+local keymap = {}
+
+--- Add a new keymap.
+--- Examples:
+--- <pre>
+---   -- You can make keymaps with Lua functions
+---   vim.keymap.set('n', 'lhs', function() print("real lua function") end)
+---
+---   -- Can use it to map multiple modes
+---   vim.keymap.set({'n', 'v'}, '<leader>lr', vim.lsp.buf.references, { silent = true, buffer=true })
+---
+---   -- You can map a key for specific buffer
+---   vim.keymap.set('n', '<leader>w', "<cmd>w<cr>", { silent = true, buffer = 5 })
+---
+---   -- Also expr keymaps too. Notice the replace_termcodes when making Lua expr maps.
+---   vim.keymap.set('i', '<Tab>', function()
+---     return vim.fn.pumvisible() == 1 and "<C-n>" or "<Tab>"
+---   end, { silent = true, expr = true, replace_termcodes = true })
+---   -- While setting <Plug> mapping remember to set remap to true
+---   vim.keymap.set('n', '[%', '<Plug>(MatchitNormalMultiBackward)', {silent = true, remap = true })
+--- </pre>
+---
+--- A pit fall of this is if you want to make keymap like:
+--- <pre>
+---    vim.keymap.set('n', 'asdf', require('jkl').my_fun)
+--- </pre>
+---
+--- `require('jkl)` get's evaluated durimg this call. If you want to
+--- avoid this cost at startup you can wrap it in a function, for example:
+--- <pre>
+---    vim.keymap.set('n', 'asdf', function() return require('jkl').my_fun() end)
+--- </pre>
+---
+---@param mode string|table   Same mode short names as |nvim_set_keymap()|.
+---                            Can also be list of modes to create keymap on multiple modes.
+---@param lhs string          Left-hand-side |{lhs}| of the mapping.
+---@param rhs string|function  Right-hand-side |{rhs}| of the mapping. Can also be a Lua function.
+--
+---@param opts table A table of |:map-arguments| such as "silent". In addition to the options
+---                  listed in |nvim_set_keymap()|, this table also accepts the following keys:
+---                  - replace_termcodes: (boolean, default false) when "true",
+---                  |nvim_replace_termcodes()| is applied to the result of Lua expr maps.
+---                  - remap: (boolean, default false) Make the mapping recursive. This is the
+---                  inverse of the "noremap" option from |nvim_set_keymap()|.
+---@see |nvim_set_keymap()|
+function keymap.set(mode, lhs, rhs, opts)
+  vim.validate {
+    mode = {mode, {'s', 't'}},
+    lhs = {lhs, 's'},
+    rhs = {rhs, {'s', 'f'}},
+    opts = {opts, 't', true}
+  }
+
+  opts = opts or {}
+  local buffer = false
+  local is_rhs_lua = type(rhs) == "function"
+  mode = type(mode) == 'string' and {mode} or mode
+
+  if opts.buffer ~= nil then
+    buffer = opts.buffer == true and 0 or opts.buffer
+    opts.buffer = nil
+  end
+
+  if opts.replace_termcodes ~= nil then
```

Comment:
My preference is this should be on by default, but I haven't read all the comments yet, so perhaps this has already been discussed.

Reply:
We haven't discussed it here . Initially I had [asked](https://github.com/neovim/neovim/pull/16594#discussion_r766309821) bfredl about applying replace termcode always on return of expr maps . But then it was declined . But I think making this default to true can be done . If someone doesn't want it they can set it to false . As far as I can see expr mapping almost always needs replace_termcodes

Reply:
+1 for making it the default from me too

Reply:
>  Initially I had asked bfredl about applying replace termcode always on return of expr maps . But then it was declined .

This was for the underlying Neovim API; I believe the Lua API is allowed to be (slightly) more magic. @bfredl?